### ISSUE TITLE
Generator objects are defined using `PyObject_HEAD` rather than `PyObject_VAR_HEAD`

### ISSUE THREAD

**Issue by jbower-fb** [python/cpython#126596]
# Bug report

### Bug description:

#29891 converted the generator objects to be variable in size so they could incorporate interpreter frames. However, the object layout is still defined using `PyObject_HEAD`. Technically I think this is incorrect. In practice probably nobody notices because the generator type is not a base class and there are other ways to work out the variable sized part where needed, so `ob_size` is never read after it is written as part of `make_gen()`.

Long story short, I want to fix this because I would like to use `ob_size` to find the size of generator objects as part of work on Meta's JIT.

### CPython versions tested on:

CPython main branch

### Linked PRs
* gh-126597

**Comment by jbower-fb** [comment]
cc @markshannon who authored https://github.com/python/cpython/pull/29891.

**Comment by markshannon** [comment]
[The PyVarObject docs](https://docs.python.org/3/c-api/structures.html#c.PyVarObject) state that "This is only used for objects that have some notion of length".
Code objects have no notion of length.

What advantage would there be to changing to `PyVarObject`?
It wouldn't provide any new capabilities, but it would use an additional 8 bytes per generator.

**Comment by jbower-fb** [comment]
Primarily it would allow me to stuff JIT-related data at the end of the generator object and not have to a bunch of work each time to calculate where that starts. I do appreciate that may not be so as useful to anyone else though so I'll see if I can find another way to mitigate the cost.

I am still left wondering though, if these objects don't have a notion of length then why don't we set `tp_itemsize` to 0 in the generator types?

**Comment by markshannon** [comment]
> why don't we set tp_itemsize to 0 in the generator types?

Because `tp_itemsize` is to tell the allocation function(s) how big the object is. Everything has a size, not everything has a length.
